oauth2_http/java/com/google/auth/oauth2/JwtCredentials.java

```diff
@@ -0,0 +1,334 @@
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.auth.oauth2;
+
+import com.google.api.client.json.JsonFactory;
+import com.google.api.client.json.webtoken.JsonWebSignature;
+import com.google.api.client.json.webtoken.JsonWebToken;
+import com.google.api.client.util.Clock;
+import com.google.auth.Credentials;
+import com.google.auth.http.AuthHttpConstants;
+import com.google.auto.value.AutoValue;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import java.io.IOException;
+import java.io.Serializable;
+import java.net.URI;
+import java.security.GeneralSecurityException;
+import java.security.PrivateKey;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * Credentials class for calling Google APIs using a JWT with custom claims.
+ *
+ * <p>Uses a JSON Web Token (JWT) directly in the request metadata to provide authorization.
+ *
+ * <pre>
+ * <code>
+ * JwtCredentials.Claims claims = JwtCredentials.Claims.newBuilder()
+ *     .setAudience("https://example.com/some-audience")
+ *     .setIssuer("some-issuer@example.com")
+ *     .setSubject("some-subject@example.com")
+ *     .build();
+ * Credentials = JwtCredentials.newBuilder()
+ *     .setPrivateKey(privateKey)
+ *     .setPrivateKeyId("private-key-id")
+ *     .setClaims(claims)
+ *     .build();
+ * </code>
+ * </pre>
+ */
+public class JwtCredentials extends Credentials implements JwtProvider {
+  private static final String JWT_ACCESS_PREFIX = OAuth2Utils.BEARER_PREFIX;
+  private static final String JWT_INCOMPLETE_ERROR_MESSAGE = "JWT claims must contain audience, "
+      + "issuer, and subject.";
+  private static final long CLOCK_SKEW = TimeUnit.MINUTES.toSeconds(5);
+
+  // byte[] is serializable, so the lock variable can be final
+  private final Object lock = new byte[0];
+  private final PrivateKey privateKey;
+  private final String privateKeyId;
+  private final Claims claims;
+  private final Long lifeSpanSeconds;
+  @VisibleForTesting
+  transient Clock clock;
+
+  private transient String jwt;
+  private transient Long expiry;
```

Comment:
error prone; use expiryInSeconds or expiryInMilliseconds as needed. Maybe make an int. Why does this need to be an object instead of a primitive?

Reply:
Will change to `expiryInSeconds` and add a comment that it's the seconds since the epoch, not how many seconds the token is good until.

For being an object, I was using null to determine whether the expire time has been set yet rather than checking against 0.